loader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/BadRow.scala

```diff
@@ -12,52 +12,142 @@
+  def compact: String  = getSelfDescribingData.asJson.noSpaces
 }
 
 object BadRow {
-  private val Line = "line"
-  private val Errors = "errors"
-
-  implicit val encoder: Encoder[BadRow] =
-    Encoder.instance { badRow: BadRow =>
-      val encoded = Base64.getEncoder.encode(badRow.line.getBytes)
-      Json.fromFields(List(
-        (Line, Json.fromString(new String(encoded))),
-        (Errors, Json.fromValues(badRow.errors.toList.map(Json.fromString)))
-      ))
-    }
+  val BQLoaderProcessorInfo = ProcessorInfo.Processor(ProjectMetadata.name, ProjectMetadata.version)
+
+  /**
+    * Represents the failure case where data can not be parsed as a proper event
+    * @param payload data blob tried to be parsed
+    * @param errors  errors in the end of the parsing process
+    */
+  final case class ParsingError(payload: String, errors: NonEmptyList[String]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.LoaderParsingError, (this: BadRow).asJson)
+  }
+
+  /**
+    * Represents the errors due to something went wrong internally
+    * In this kind of errors, event is parsed properly, therefore
+    * enriched event is given in the error as properly parsed
+    * @param enrichedEvent event which is enriched successfully
+    * @param errors        info of errors
+    */
+  final case class InternalError(enrichedEvent: Event, errors: NonEmptyList[InternalErrorInfo]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.LoaderInternalError, (this: BadRow).asJson)
+  }
 
-  implicit val decoder: Decoder[BadRow] =
-    Decoder.instance[BadRow] { cursor =>
-      for {
-        jsonObject <- cursor.value
-          .asObject
-          .toRight(DecodingFailure("Bad row should be an object", cursor.history))
-        map = jsonObject.toMap
-        line <- map.get(Line)
-          .flatMap(_.asString)
-          .toRight(DecodingFailure(s"$Line should be a string", cursor.history))
-        decodedLine <- Either.catchNonFatal(new String(Base64.getDecoder.decode(line)))
-          .leftMap(error => DecodingFailure(s"Cannot decode $Line ${error.getMessage}", cursor.history))
-        errors <- map.get(Errors)
-          .flatMap(_.as[NonEmptyList[String]].toOption)
-          .toRight(DecodingFailure(s"Cannot decode $Errors", cursor.history))
-
-      } yield BadRow(decodedLine, errors)
+  implicit val badRowCirceJsonEncoder: Encoder[BadRow] =
+    Encoder.instance {
+      case ParsingError(payload, errors) => json"""{
+          "payload": $payload,
+          "errors": $errors,
+          "processor": $BQLoaderProcessorInfo
+        }"""
+      case InternalError(enrichedEvent, errors) => json"""{
+          "enrichedEvent": $enrichedEvent,
+          "failures": $errors,
+          "processor": $BQLoaderProcessorInfo
+        }"""
     }
+
+  /**
+    * Gives info about the reasons of the internal errors
```

Comment:
I think what we need here is exmplanation what is "internal" error.

Reply:
I think it would be better to remove all `internal` keyword usage. I will write more detailed comment than.